### ISSUE TITLE
http2 client support?

### ISSUE THREAD

**Issue by c0b** [googleapis/google-cloud-node#2616]
since most gcloud service support http2 already, and nodejs v8.5 now has http2 support https://nodejs.org/api/http2.html do you have plans to support nodejs client side also in http2?

my application has many concurrent bigquery calls, in `lsof -nP -p <pid>` output right now, I see it creates a new tcp connection for each bigquery API call, if have http2 client support, then all these connections can be consolidated into one http2 connection and do multiplexing ?

```console
COMMAND   PID USER   FD      TYPE DEVICE SIZE/OFF    NODE NAME
node    12662  drc   12u     IPv4 139859      0t0     TCP 10.42.0.129:48242->172.217.5.74:443 (ESTABLISHED)
node    12662  drc   13u     IPv4 140086      0t0     TCP 10.42.0.129:48710->172.217.5.74:443 (ESTABLISHED)
node    12662  drc   15u     IPv4 140083      0t0     TCP 10.42.0.129:48704->172.217.5.74:443 (ESTABLISHED)
node    12662  drc   16u     IPv4 137916      0t0     TCP 10.42.0.129:48682->172.217.5.74:443 (ESTABLISHED)
node    12662  drc   17u     IPv4 139861      0t0     TCP 10.42.0.129:48246->172.217.5.74:443 (ESTABLISHED)
node    12662  drc   18u     IPv4 139862      0t0     TCP 10.42.0.129:48248->172.217.5.74:443 (ESTABLISHED)
node    12662  drc   19u     IPv4 139863      0t0     TCP 10.42.0.129:48250->172.217.5.74:443 (ESTABLISHED)
node    12662  drc   20u     IPv4 139864      0t0     TCP 10.42.0.129:48252->172.217.5.74:443 (ESTABLISHED)
node    12662  drc   21u     IPv4 139865      0t0     TCP 10.42.0.129:48254->172.217.5.74:443 (ESTABLISHED)
node    12662  drc   22u     IPv4 139866      0t0     TCP 10.42.0.129:48256->172.217.5.74:443 (ESTABLISHED)
node    12662  drc   23u     IPv4 139867      0t0     TCP 10.42.0.129:48258->172.217.5.74:443 (ESTABLISHED)
node    12662  drc   24u     IPv4 139868      0t0     TCP 10.42.0.129:48260->172.217.5.74:443 (ESTABLISHED)
node    12662  drc   25u     IPv4 139869      0t0     TCP 10.42.0.129:48262->172.217.5.74:443 (ESTABLISHED)
node    12662  drc   26u     IPv4 139870      0t0     TCP 10.42.0.129:48264->172.217.5.74:443 (ESTABLISHED)
node    12662  drc   27u     IPv4 139871      0t0     TCP 10.42.0.129:48266->172.217.5.74:443 (ESTABLISHED)
node    12662  drc   28u     IPv4 139872      0t0     TCP 10.42.0.129:48268->172.217.5.74:443 (ESTABLISHED)
```

**Comment by stephenplusplus** [comment]
@ofrobots @lukesneeringer do you know of any plans to make this shift towards http2?

**Comment by lukesneeringer** [comment]
For gRPC libraries, yes. For REST libraries, I am not sure. I am definitely _willing_, but time is another issue.

One open question is whether h2 is going to be backported to Node 6. If it is (and my understanding is that this is the plan), then we could add h2 support without multiplexing (e.g. just change over wholesale)  in April 2018. That would be a significantly easier lift. If it is not backported, then we have to multiplex until Node 6 goes EOL in April 2019.

Short version: I would like to do it! But it is not on the immediate roadmap at this time.

**Comment by stephenplusplus** [comment]
I'll call this blocked now:

- GRPC/GAX APIs: Pending gRPC support (ETA: ["a few months"](https://github.com/GoogleCloudPlatform/google-cloud-node/issues/2697#issuecomment-340082210) as of September 2017)
- REST APIs: Pending http2 support in Node v6. When Node@4 goes EOL, April 2018

**Comment by ofrobots** [comment]
For REST, an alternative would be if we swapped `request` for another module that would transparently use h2 transport when available and use h1 otherwise. We independently would like to drop `request` anyway, but I am not sure if there is a viable alternative yet. /cc @jmdobry

**Comment by c0b** [comment]
the [`node-fetch`](https://www.npmjs.com/package/node-fetch) is trying to bring the browser standard `fetch` API into Node and be transparent to h2 and h1 servers while that could be hard because of nodejs/node#16256 not done yet
the [`fetch-h2`](https://www.npmjs.com/package/fetch-h2) is focusing on h2 servers, I feel this could be the one you can replace `request` here @ofrobots however for Nodejs version 8+ users only it's hard to support Node v6 users in the same way.

**Comment by bcoe** [comment]
There are now gRPC clients available for most GCP APIs. There is also the option to use an HTTP2 transport for google-api-nodejs-client.
